### ISSUE TITLE
Suggestions: Limiting results to increase performance and caching discussion

### ISSUE THREAD

**Issue by pgrunm-DrKlein** [celestialorb/aws-ecr-prometheus-exporter#3]
Hello,

it's me again😄. I'd like to suggest or discuss some things

- How do you plan to implement caching? Any repository which the exporter retrieves is caches for an amount x? Also the findings etc? What about cache saving then?
- Performance: To increase performance I would like to suggest to only retrieve the latest x images of a repo. This would lead to both an increase of performance as well decreasing the chance to hit rate limits. Make this configurable with an environment variable -> ever better.

**Comment by celestialorb** [comment]
I was planning to implement just a simple in-memory cache for the security findings as those are only updated once every twenty-four hours anyway. There's no need to call the API unless that cache is expired; so on start it will call the API for all security findings and then only once every twenty-four hours (unless your process restarts, in which case it will then start hitting the API again).

The cache doesn't really need to store the results of the call, just that the call was made as the results are effectively stored via the Prometheus metrics that are exported.

This could be improved with the use of an external cache like Redis, but I'd like to avoid adding a dependency for a simple Prometheus exporter unless we really need it. An external cache would avoid calling the API again on process restart.

Even with the cache this still hits rate-limiting issues, even with client-side rate limiting. So the only other improvement I can think of is, as you mention, limiting the number of images we grab security findings/metrics for. This would of course be configurable with maybe a sane default (not sure if we should have a default limit, or leave it unlimited by default).

In my use case I'd configure it to be unlimited as I want to know the security findings of all images, as in my environment all images in ECR represent a use in our cluster somewhere (i.e. old images that are no longer used are deleted after some time).

**Comment by celestialorb** [comment]
There's also the option of disabling collection of certain metrics if a user isn't interested in them. If the security findings metric calls are too heavy for a user and they're not particularly interested in them they could have the option to disable that collection entirely.

**Comment by pgrunm-DrKlein** [comment]
I think you're right with the cache, there is no need for external dependencies like Redis. If we can keep this simple then why not.

About the number of images I did some quick maths. We currently have about 3500 different images and scraping every single one would take like forever. If the rate limit allows one call per second this would be almost an hour to fetch all of them. Like you said, even then it's likely to hit rate limits, so maybe a default value in between 10 and unlimited might be useful. For example our average image count per repo is 34. The chance is high other organizations also have a few repos with a high number of images and more repos with less images. I prefer keeping this setting lower like 10 or 25 so you don't run in this rate limiting issue when using default parameters. 

Another thing to keep in mind is the more images you scrape, the more metrics you export and this can also result in more cost because of increased traffic when scraping in case you're using a cloud provider like AWS. 

Two more things came to my mind when I wrote this wall of text:

- I'd love a metric to see how long the scraping took, maybe a gauge?
- If scraping all metrics can take up to several hours, we should publish the metric as soon as the image and findings was parsed. 👍

**Comment by pgrunm-DrKlein** [comment]
If I can you help you with development, please tell me.

**Comment by celestialorb** [comment]
@pgrunm-DrKlein Sorry, I was out of town at a conference last week. I released v0.3.0 during that time that adds in limiting the number of tags scanned per image: https://github.com/celestialorb/aws-ecr-prometheus-exporter/releases/tag/v0.3.0.

The additional configuration is in the [README](https://github.com/celestialorb/aws-ecr-prometheus-exporter#usage), but it comes with a limit setting as well as a toggle though it is disabled by default so you'll need to explicitly enable it.

**Comment by pgrunm-DrKlein** [comment]
Yes, I saw that. :) I just wanted to assist you with development, because I really appreciate your work. I think this project is really useful and surely takes so time to develop. 👍
